// components/Articles/Post/index.js
import React, { Component } from 'react';

import { actions } from '_firebase';
import style from '../articles.css';


import FontAwesome from 'react-fontawesome';
import Loading from 'components/Loading';
import Header from './header';

class NewsArticle extends Component {

  state = {
    fetching: true,
    article: null,
    team: null,
    image: ''
  }

  componentDidMount = async () => {
    let res = await actions.findArticleById('articles', this.props.match.params.id).catch(e => this.props.history.push('/'));
    if (!res) return;

    let {article, team} = res;
    this.setState({article, team, fetching: false});
    actions.getImageUrl(article.image)
      .then( url => this.setState({ image: url }) );
    
  }

  handleDeleteBtn = (e) => {
    this.setState({deleting: true});
    actions
      .deleteArticle('articles', this.state.article)
      .then(() => {
        this.props.history.push('/');
      });
  }

  renderDeleteButton = () => (
    (this.state.article && actions.confirmUser(this.state.article.author_id)) ?
      <div className={style.deleteBtn} onClick={this.handleDeleteBtn}>
        {
          this.state.deleting ?
            <Loading/>
          :
            <FontAwesome name="trash-alt"/>
        }
      </div>
    :
      null
  )

  renderBody = (title, text, image) => (
    <div className={style.articleBody}>
      <h1>{title}</h1>
      <div className={style.articleImage} style={{backgroundImage: `url('${this.state.image}')`}}></div>
      <div
        className={style.articleText}
        dangerouslySetInnerHTML={{
          __html: text
        }}

      >
      </div>
    </div>
  )

  render() {

    let {article, team} = this.state;
    if (!article || !team) return null;

    return (
      this.state.fetching ?
        <div className={style.loading}><Loading/></div>
      :
        <div className={style.wrapper}>
          <Header
            team={team}
            date={article.date}
            author={article.author}
          />
          {this.renderBody(article.title, article.body, article.image)}
          {this.renderDeleteButton()}
        </div>
    );
  }
}

export default NewsArticle;
