// components/widgets/NewsSlider/newsSlider.js
import React, { Component } from 'react';

import SliderTemplate from './sliderTemplate';
import { actions } from '_firebase';
import Empty from '../Empty';
import Loading from '../../Loading';

class NewsSlider extends Component {

  state = {
    news: [],
    start: this.props.start,
    amount: this.props.amount,
    fetching: true
  }

  request = async (start) => {
    let news = await actions.findArticles('articles', this.state.start, this.state.amount);
    start = this.state.start + this.state.amount;

    this.setState({
      news: this.state.news.concat(news), 
      start,
      fetching: false
    });

  }

  componentDidMount() {
    this.request(this.state.start);
  }

  renderBody = () => (
    this.state.fetching ? 
      <div style={{padding: '10px', textAlign: 'center'}}><Loading/></div>
    :
      !this.state.news.length && !this.state.fetching ? 
        <Empty/>
      :
        <SliderTemplate
          data={this.state.news}
          type={this.props.type}
        />
  )

  render() {

    return this.renderBody();
  }
}

export default NewsSlider;
